libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -1229,623 +1230,1733 @@ private static (object, string) ParseISOTimestamp(string timeStamp, Func<DateTim
             return (result, error);
         }
 
-        private static string AddOrdinal(int num)
+        private static (string, string) ConvertFromUTC(string utcTimestamp, string timezone, string format = "o")
         {
-            var hasResult = false;
-            var ordinalResult = num.ToString();
-            if (num > 0)
+            string error = null;
+            string result = null;
+            TimeZoneInfo tz = null;
+            var utcDt = DateTime.UtcNow;
+            try
             {
-                switch (num % 100)
+                utcDt = DateTime.Parse(utcTimestamp).ToUniversalTime();
+            }
+            catch
+            {
+                error = "illegal timestamp format";
+            }
+
+            if (error == null)
+            {
+                try
                 {
-                    case 11:
-                    case 12:
-                    case 13:
-                        ordinalResult += "th";
-                        hasResult = true;
-                        break;
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                    {
+                        var convertedTZ = TimeZoneConverter.WindowsToIana(timezone);
+                        tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);
+                    }
+                    else
+                    {
+                        var convertedTZ = TimeZoneConverter.IanaToWindows(timezone);
+                        tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);
+                    }
+                }
+                catch
+                {
+                    error = "illegal timezone info";
```

Comment:
error message could include more info such as $"{timezone} is an illegal timezone"